Mathlib/Geometry/Manifold/MeasureZero.lean

```diff
@@ -0,0 +1,163 @@
+/-
+Copyright (c) 2023 Michael Rothgang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Rothgang
+-/
+import Mathlib.Geometry.Manifold.SmoothManifoldWithCorners
+import Mathlib.Topology.MetricSpace.Polish
+import Mathlib.MeasureTheory.Measure.Haar.Basic
+
+/-!
+# Measure zero subsets of a manifold
+
+Defines the notion of "measure zero" subsets on a finite-dimensional topological manifold.
+
+Let $M$ be a finite-dimensional manifold. While $M$ is a measurable space
+(for instance, pull back the Lebesgue measure on each chart and use a partition of unity),
+it has no canonical measure. However, there is a natural notion of *measure zero* subsets of $M$:
+a subset $A\subset M$ has measure zero iff for each chart $(\phi, U)$ of $M$ the set
+$\phi(U\cap A)\subset R^n$ has measure zero (w.r.t. to any additive Haar measure).
+
+Measure zero sets are closed under subsets and countable unions, hence their complement defines
+the **almost everywhere filter** on a manifold. Measure zero subsets have empty interior,
+hence closed measure zero sets are nowhere dense.
+
+**TODO**
+- show that if $M$ is a normed space with Haar measure $\mu$,
+  a set $A ⊆ M$ has measure zero iff $μ A = 0$.
+- show the same if $M$ is a manifold with a suitable measure $\mu$. If `MeasureZero` were defined
+using `IsOpenPosMeasure` on a chart, "suitable" probably would mean that.
+- include manifolds with boundary: in `open_implies_empty`, one needs to show that
+an open subset of `M` includes an interior point of `M`. (The current definition of
+manifolds with boundary seems to be too general for that.)
+
+
+## Main results
+- `MeasureZero`: Prop for measure zero subsets in `M`
+- `MeasureZero.mono`: measure zero subsets are closed under subsets
+- `MeasureZero.union`, `MeasureZero.iUnion`: measure zero subsets are closed under countable unions
+- `MeasureZero.ae`: the complements of measure zero sets form the **almost everywhere filter**.
+- `open_implies_empty`: an open measure zero subset is empty
+- `MeasureZero_implies_empty_interior`: a measure zero subset has empty interior.
+
+## References
+See [Hirsch, Differential Topology](Hirsch76), chapter 3.1 for this definition.
+-/
+
+open MeasureTheory Measure Function TopologicalSpace Set
+set_option autoImplicit false
+
+variable
+  -- Let `M` be a finite-dimensional topological manifold without boundary over the pair `(E, H)`.
+  {E : Type*} [NormedAddCommGroup E] [NormedSpace ℝ E] {H : Type*} [TopologicalSpace H]
+  (I : ModelWithCorners ℝ E H) {M : Type*} [TopologicalSpace M] [ChartedSpace H M] [I.Boundaryless]
+  [FiniteDimensional ℝ E] [MeasurableSpace E] [BorelSpace E]
+variable {I}
```

Comment:
Why not make `I` implicit to begin with?

Reply:
`I` must be explicit in the definition of measure zero sets (otherwise, typeclass inference gets stuck several times).

Reply:
Yes, but you already use `variable (I) in` for that that definition. In this case, my question is why usesure that
```lean
variable ... (I : ModelWithCorners ℝ E H) ...
variable {I}
```
instead of
```lean
variable ... {I : ModelWithCorners ℝ E H} ...
```

Reply:
Ah, that: there's no good reason; I didn't think about this.